### PR TITLE
Add update count test in e2e/pods.go

### PR DESCRIPTION
**Pull request by Random-Liu**
In this PR, I added update count test in e2e/pods.go, so as to help us understand pod updates during single pod creation and deletion. Notice that this test only counts update number but hasn't set any upper bound now.

Originally, this is also part of #18410. I split this to a separate PR to make it easier for reviewing.

@yujuhong

### CODE REVIEW

File: test/e2e/pods.go [hunk 1]
```diff
@@ -466,6 +469,116 @@ var _ = Describe("Pods", func() {
 		Logf("Pod update OK")
 	})
 
+	It("should have proper update count during pod creation and deletion", func() {
+		// This e2e test is mainly aimed at understand the update count for pod creation and deletion
```
Comment (yujuhong) [inline test/e2e/pods.go]: s/This e2e test is mainly aimed at understand/This test aims to understand

File: test/e2e/pods.go [hunk 2]
```diff
@@ -466,6 +469,116 @@ var _ = Describe("Pods", func() {
...
+		// TODO(random-liu): Limit the update count later when we have a reasonable upper bound.
+		podClient := framework.Client.Pods(framework.Namespace.Name)
+
+		By("creating the pod")
```
Comment (yujuhong) [inline test/e2e/pods.go]: You don't actually create the pod (through the apiserver) here. The description is a little misleading.

File: test/e2e/pods.go [hunk 3]
```diff
@@ -466,6 +469,116 @@ var _ = Describe("Pods", func() {
...
+					updates++
+				},
+				DeleteFunc: func(_ interface{}) {
+					updates++
+				},
+			},
+		)
+		go controller.Run(stop)
+
+		By("wait until stable")
```
Comment (yujuhong) [inline test/e2e/pods.go]: Why waiting until stable here? The pod hasn't even been created yet.

File: test/e2e/pods.go (hunk 1 above)
Comment (Random-Liu) [inline test/e2e/pods.go]: Thanks! :)

File: test/e2e/pods.go [hunk 4]
```diff
@@ -466,6 +469,116 @@ var _ = Describe("Pods", func() {
...
+		// the test so we can ensure that we clean up after
+		// ourselves
+		defer podClient.Delete(pod.Name, api.NewDeleteOptions(0))
+		pod, err = podClient.Create(pod)
+		if err != nil {
+			Failf("Failed to create pod: %v", err)
+		}
+		By("wait until stable")
+		waitFunc()
+		current := updates
```
Comment (yujuhong) [inline test/e2e/pods.go]: `updates` is accessed by multiple goroutines. There will be a race.

File: test/e2e/pods.go [hunk 5]
```diff
@@ -466,6 +469,116 @@ var _ = Describe("Pods", func() {
...
+				Labels: map[string]string{
+					"name": name,
+					"time": value,
+				},
+			},
+			Spec: api.PodSpec{
+				Containers: []api.Container{
+					{
+						Name:  "nginx",
+						Image: "gcr.io/google_containers/nginx:1.7.9",
```
Comment (yujuhong) [inline test/e2e/pods.go]: Why use nginx? Does this add something to the test?

File: test/e2e/pods.go (hunk 2 above)
Comment (Random-Liu) [inline test/e2e/pods.go]: Yeah, it's really a little misleading...But I copied this from the other test cases.

File: test/e2e/pods.go [hunk 6]
```diff
@@ -466,6 +469,116 @@ var _ = Describe("Pods", func() {
...
+		value := strconv.Itoa(time.Now().Nanosecond())
+		pod := &api.Pod{
+			ObjectMeta: api.ObjectMeta{
+				Name: name,
+				Labels: map[string]string{
+					"name": name,
+					"time": value,
+				},
+			},
+			Spec: api.PodSpec{
```
Comment (yujuhong) [inline test/e2e/pods.go]: nit: explain why you use this container image with the liveness probe.

File: test/e2e/pods.go [hunk 7]
```diff
@@ -466,6 +469,116 @@ var _ = Describe("Pods", func() {
...
+
+		By("deleting the pod gracefully")
+		if err := podClient.Delete(pod.Name, nil); err != nil {
+			Failf("Failed to delete pod: %v", err)
+		}
+		waitFunc()
+		current = updates
+		Logf("Pod Deletion Update Count: %d", current-last)
+
+		close(stop)
```
Comment (yujuhong) [inline test/e2e/pods.go]: you can defer channel closing as well.

File: test/e2e/pods.go (hunk 3 above)
Comment (Random-Liu) [inline test/e2e/pods.go]: Yeah, I should remove this. I want to make sure there is no other updates on the way before, but now I only listen to the specific pod, I should remove it. Thanks.

File: test/e2e/pods.go (hunk 7 above)
Comment (Random-Liu) [inline test/e2e/pods.go]: ACK

**Comment by yujuhong** [comment]
I am ok with this test, but unless this test is run in serial, and you can make sure there are no pending updates from the node, the number of updates may vary a lot (due to our `syncBatch` mechanism).

File: test/e2e/pods.go (hunk 4 above)
Comment (Random-Liu) [inline test/e2e/pods.go]: I borrowed the idea from density test. There will be race, but after wait until stable, the chance of race is really slim. I can add a lock if you think it is necessary, :)

File: test/e2e/pods.go (hunk 6 above)
Comment (Random-Liu) [inline test/e2e/pods.go]: ACK...I just use the same pod spec with the other test, but I think I should simplify it here. Thanks.

**Comment by Random-Liu** [comment]
Yeah, this test was written for helping us find the best implementation for reconciliation, in most cases I run it separately.
Because now it doesn't do any meaningful check, I can:
* Keep it in my local repo, and re-implement it when it is meaningful in the future.
* Skip it.

Either is OK for me, ;)

File: test/e2e/pods.go (hunk 4 above)
Comment (yujuhong) [inline test/e2e/pods.go]: I believe density test closes the channel to stop the goroutine before checking the update. A potential race in the code is problematic. You can either add a lock or use the golang atomic package  :-)

**Comment by Random-Liu** [comment]
Close this PR for now, because it is not that meaningful now.

**Pull request closed without merging**